compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -636,6 +636,33 @@ public void visit(BLangBinaryExpr binaryExpr) {
         resultType = types.checkType(binaryExpr, actualType, expType);
     }
 
+    private BSymbol getBinaryEqualityForTypeSets(OperatorKind opKind, BType lhsType,
+                                                 BType rhsType, BLangBinaryExpr binaryExpr) {
+        if (opKind != OperatorKind.EQUAL && opKind != OperatorKind.NOT_EQUAL) {
+            return symTable.notFoundSymbol;
+        }
+        if (types.isIntersectionExist(lhsType, rhsType)) {
+            if ((!types.isValueType(lhsType) && !types.isValueType(rhsType)) ||
+                    (types.isValueType(lhsType) && types.isValueType(rhsType))) {
+                int opcode = (opKind == OperatorKind.EQUAL) ? InstructionCodes.REQ : InstructionCodes.RNE;
+                List<BType> paramTypes = Lists.of(lhsType, rhsType);
+                BType retType = symTable.booleanType;
+                BInvokableType opType = new BInvokableType(paramTypes, retType, null);
+                return new BOperatorSymbol(names.fromString(opKind.value()), null, opType, null, opcode);
```

Comment:
Shall we move this symbol creation logic to SymbolResolver?
Same can be re-used in the `else` block.

Reply:
I also thought the same, for the second else case here will require a implicit cast which will require the expression, not only just expression type, similarly cases we resolve symbols in resolver. So I kept the code here.